libsrc/leddevice/dev_ftdi/ProviderFtdi.cpp

```diff
@@ -0,0 +1,248 @@
+		setInError(ftdi_get_error_string(_ftdic));
+		return rc;
+	}
+
+	double reference_clock = 60e6;
+	int divisor = (reference_clock / 2 / _baudRate_Hz) - 1;
+	uint8_t buf[10] = {0};
+	unsigned int icmd = 0;
+	buf[icmd++] = DIS_DIV_5;
+	buf[icmd++] = TCK_DIVISOR;
+	buf[icmd++] = divisor;
+	buf[icmd++] = divisor >> 8;
+	buf[icmd++] = SET_BITS_LOW;		  // opcode: set low bits (ADBUS[0-7])
+	buf[icmd++] = Pin::CS & ~Pin::L0; // argument: inital pin states
+	buf[icmd++] = pinDirection;
+	if ((rc = ftdi_write_data(_ftdic, buf, icmd)) != icmd)
+	{
+		setInError(ftdi_get_error_string(_ftdic));
+		return rc;
+	}
+
+	_isDeviceReady = true;
+	return rc;
+}
+
+int ProviderFtdi::close()
+{
+	// allow to clock out remaining data from powerOff()->writeBlack()
+	if (_ftdic != NULL)
+	{
+		std::this_thread::sleep_for(std::chrono::milliseconds(15));
+		Debug(_log, "Closing FTDI device");
+		ftdi_set_bitmode(_ftdic, 0x00, BITMODE_RESET);
+		ftdi_usb_close(_ftdic);
+		_ftdic = NULL;
+	}
+	return LedDevice::close();
+}
+
+void ProviderFtdi::setInError(const QString &errorMsg, bool isRecoverable)
+{
+	close();
+
+	LedDevice::setInError(errorMsg);
+}
+
+int ProviderFtdi::writeBytes(const qint64 size, const uint8_t *data)
+{
+	uint8_t buf[10] = {0};
+	unsigned int icmd = 0;
+	int rc = 0;
+
+	int count_arg = size - 1;
+	buf[icmd++] = SET_BITS_LOW;
+	buf[icmd++] = Pin::L0 & ~Pin::CS;
+	buf[icmd++] = pinDirection;
+	buf[icmd++] = MPSSE_DO_WRITE | MPSSE_WRITE_NEG;
+	buf[icmd++] = count_arg;
+	buf[icmd++] = count_arg >> 8;
+
+	if ((rc = ftdi_write_data(_ftdic, buf, icmd)) != icmd)
+	{
+		setInError(ftdi_get_error_string(_ftdic));
+		return rc;
+	}
+	if ((rc = ftdi_write_data(_ftdic, data, size)) != size)
+	{
+		setInError(ftdi_get_error_string(_ftdic));
+		return rc;
+	}
+	icmd = 0;
+	buf[icmd++] = SET_BITS_LOW;
+	buf[icmd++] = Pin::CS & ~Pin::L0;
+	buf[icmd++] = pinDirection;
+	if ((rc = ftdi_write_data(_ftdic, buf, icmd)) != icmd)
+	{
+		setInError(ftdi_get_error_string(_ftdic));
+		return rc;
+	}
+	return rc;
+}
+
+QJsonObject ProviderFtdi::discover(const QJsonObject & /*params*/)
+{
+	QJsonObject devicesDiscovered;
+	QJsonArray deviceList;
+	struct ftdi_device_list *devlist;
+	struct ftdi_context *ftdic;
+
+	QJsonObject autoDevice = QJsonObject{{"value", AUTO_SETTING}, {"name", "Auto"}};
+	deviceList.push_back(autoDevice);
+
+	ftdic = ftdi_new();
+
+	if (ftdi_usb_find_all(ftdic, &devlist, ANY_FTDI_VENDOR, ANY_FTDI_PRODUCT) > 0)
+	{
+		struct ftdi_device_list *curdev = devlist;
+		while (curdev)
+		{
+			char manufacturer[128], description[128];
+			ftdi_usb_get_strings(ftdic, curdev->dev, manufacturer, 128, description, 128, NULL, 0);
+
+			libusb_device_descriptor desc;
+			libusb_get_device_descriptor(curdev->dev, &desc);
+			QString value = QString("i:0x%1:0x%2")
+								.arg(desc.idVendor, 4, 16, QChar{'0'})
+								.arg(desc.idProduct, 4, 16, QChar{'0'});
+
+			QString name = QString("%1 (%2)").arg(manufacturer, description);
+			deviceList.push_back(QJsonObject{
```

Comment:
You might want to break down the device info similar as for the HID devices

https://github.com/hyperion-project/hyperion.ng/blob/master/libsrc/leddevice/dev_hid/ProviderHID.cpp

The in js the "serial" section can be reused populating the device selection list.
Are udev devices for FTDI devices supported?

Is there a need to differentiate the device by its serial number, in case multiple devices are connected to the system?